<img src="http://140.116.53.170/classes/numericalOptimization/simplexMethods/figure001.png" width="600" alt="Given Information">

In [1]:
import numpy as np
def E1(A,i,k):
    m=len(A)
    n=len(A[0,:])
    I=np.identity(m)
    I[i,i]=k
    return I.dot(A)
def E2(A,i,j,k):
    m=len(A)
    n=len(A[0,:])
    I=np.identity(m)
    I[j,i]=k
    return I.dot(A)
np.set_printoptions(precision=3,suppress=True)

In [3]:
A=np.zeros((4,7))
A[0,0]=1
A[0,1]=-3
A[0,2]=-5
A[0,3]=0
A[0,4]=0
A[0,5]=0
A[0,6]=0

A[1,0]=0
A[1,1]=1
A[1,2]=0
A[1,3]=1
A[1,4]=0
A[1,5]=0
A[1,6]=4

A[2,0]=0
A[2,1]=0
A[2,2]=2
A[2,3]=0
A[2,4]=1
A[2,5]=0
A[2,6]=12

A[3,0]=0
A[3,1]=3
A[3,2]=2
A[3,3]=0
A[3,4]=0
A[3,5]=1
A[3,6]=18

print(A)

m=len(A)
n=len(A[0,:])
print(m,n)

rq=-1
while rq<0: 
    #######################################
    rq=float("inf")
    j=-1
    for l in range(1,n-1):
        if A[0,l]<rq:
            rq=A[0,l]
            j=l
    print(rq,j)
    #######################################    
    if rq>=0:
        break
    #######################################
    MRT=float("inf")
    i=-1
    for l in range(1,m):
        if A[l,j]>0:
            if A[l,n-1]/A[l,j]<MRT:
                MRT=A[l,n-1]/A[l,j]
                i=l
    print(MRT,i)
    #######################################
    A=E1(A,i,1/A[i,j])
    A=E2(A,i,0,-A[0,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    #for l in range(0,i):    
    #    A=E2(A,i,0,-A[l,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    for l in range(i+1,m):
        A=E2(A,i,l,-A[l,j])
    #######################################
    print(A)

[[ 1. -3. -5.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.]
 [ 0.  0.  2.  0.  1.  0. 12.]
 [ 0.  3.  2.  0.  0.  1. 18.]]
4 7
-5.0 2
6.0 2
[[ 1.  -3.   0.   0.   2.5  0.  30. ]
 [ 0.   1.   0.   1.   0.   0.   4. ]
 [ 0.   0.   1.   0.   0.5  0.   6. ]
 [ 0.   3.   0.   0.  -1.   1.   6. ]]
-3.0 1
2.0 3
[[ 1.     0.     0.     0.     1.5    1.    36.   ]
 [ 0.     0.     0.     1.     0.333 -0.333  2.   ]
 [ 0.     0.     1.     0.     0.5    0.     6.   ]
 [ 0.     1.     0.     0.    -0.333  0.333  2.   ]]
0.0 1


In [4]:
#insert the information of minimum ratioj test into the table
A=np.zeros((4,8))
A[0,0]=1
A[0,1]=-3
A[0,2]=-5
A[0,3]=0
A[0,4]=0
A[0,5]=0
A[0,6]=0

A[1,0]=0
A[1,1]=1
A[1,2]=0
A[1,3]=1
A[1,4]=0
A[1,5]=0
A[1,6]=4

A[2,0]=0
A[2,1]=0
A[2,2]=2
A[2,3]=0
A[2,4]=1
A[2,5]=0
A[2,6]=12

A[3,0]=0
A[3,1]=3
A[3,2]=2
A[3,3]=0
A[3,4]=0
A[3,5]=1
A[3,6]=18

print(A)

m=len(A)
n=len(A[0,:])
print(m,n)

rq=-1
while rq<0: 
    #######################################
    rq=float("inf")
    j=-1
    for l in range(1,n-1):
        if A[0,l]<rq:
            rq=A[0,l]
            j=l
    print(rq,j)
    #######################################    
    if rq>=0:
        break
    #######################################
    MRT=float("inf")
    i=-1
    for l in range(1,m):
        if A[l,j]>0:
            A[l,n-1] = A[l,n-2]/A[l,j]
            if A[l,n-2]/A[l,j]<MRT:
                MRT=A[l,n-2]/A[l,j]
                i=l
    print(MRT,i)
    print(A)
    #######################################
    A=E1(A,i,1/A[i,j])
    A=E2(A,i,0,-A[0,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    #for l in range(0,i):    
    #    A=E2(A,i,0,-A[l,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    for l in range(i+1,m):
        A=E2(A,i,l,-A[l,j])
    #######################################
    for l in range(0,m):
        A[l,n-1]=0
print(A)

[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  0.]
 [ 0.  3.  2.  0.  0.  1. 18.  0.]]
4 8
-5.0 2
6.0 2
[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  6.]
 [ 0.  3.  2.  0.  0.  1. 18.  9.]]
-3.0 1
2.0 3
[[ 1.  -3.   0.   0.   2.5  0.  30.   0. ]
 [ 0.   1.   0.   1.   0.   0.   4.   4. ]
 [ 0.   0.   1.   0.   0.5  0.   6.   0. ]
 [ 0.   3.   0.   0.  -1.   1.   6.   2. ]]
0.0 1
[[ 1.     0.     0.     0.     1.5    1.    36.     0.   ]
 [ 0.     0.     0.     1.     0.333 -0.333  2.     0.   ]
 [ 0.     0.     1.     0.     0.5    0.     6.     0.   ]
 [ 0.     1.     0.     0.    -0.333  0.333  2.     0.   ]]


In [5]:
#produce a subroutine for updating the tableau whenever the tableau, ith row and jth column are given.
def updateTableau(A,i,j,m,n):
    A=E1(A,i,1/A[i,j])
    A=E2(A,i,0,-A[0,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    #for l in range(0,i):    
    #    A=E2(A,i,0,-A[l,j])
    for l in range(1,i):
        A=E2(A,i,l,-A[l,j])
    for l in range(i+1,m):
        A=E2(A,i,l,-A[l,j])
    return A

In [6]:
#produce a subroutine for generating the tableaus for all steps in the simplex method
def  generaTableaus(A):
    m=len(A)
    n=len(A[0,:])
    print(m,n)

    rq=-1
    while rq<0: 
        #######################################
        rq=float("inf")
        j=-1
        for l in range(1,n-1):
            if A[0,l]<rq:
                rq=A[0,l]
                j=l
        print(rq,j)
        #######################################    
        if rq>=0:
            break
        #######################################
        MRT=float("inf")
        i=-1
        for l in range(1,m):
            if A[l,j]>0:
                A[l,n-1] = A[l,n-2]/A[l,j]
                if A[l,n-2]/A[l,j]<MRT:
                    MRT=A[l,n-2]/A[l,j]
                    i=l
        print(MRT,i)
        print(A)
        A=updateTableau(A,i,j,m,n)
        #######################################
        for i in range(m):
            A[i,n-1]=0
    return A


In [7]:
A=np.zeros((4,8))
A[0,0]=1
A[0,1]=-3
A[0,2]=-5
A[0,3]=0
A[0,4]=0
A[0,5]=0
A[0,6]=0

A[1,0]=0
A[1,1]=1
A[1,2]=0
A[1,3]=1
A[1,4]=0
A[1,5]=0
A[1,6]=4

A[2,0]=0
A[2,1]=0
A[2,2]=2
A[2,3]=0
A[2,4]=1
A[2,5]=0
A[2,6]=12

A[3,0]=0
A[3,1]=3
A[3,2]=2
A[3,3]=0
A[3,4]=0
A[3,5]=1
A[3,6]=18

print(A)
generaTableaus(A)

[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  0.]
 [ 0.  3.  2.  0.  0.  1. 18.  0.]]
4 8
-5.0 2
6.0 2
[[ 1. -3. -5.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  1.  0.  0.  4.  0.]
 [ 0.  0.  2.  0.  1.  0. 12.  6.]
 [ 0.  3.  2.  0.  0.  1. 18.  9.]]
-3.0 1
2.0 3
[[ 1.  -3.   0.   0.   2.5  0.  30.   0. ]
 [ 0.   1.   0.   1.   0.   0.   4.   4. ]
 [ 0.   0.   1.   0.   0.5  0.   6.   0. ]
 [ 0.   3.   0.   0.  -1.   1.   6.   2. ]]
0.0 1


array([[ 1.   ,  0.   ,  0.   ,  0.   ,  1.5  ,  1.   , 36.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  1.   ,  0.333, -0.333,  2.   ,  0.   ],
       [ 0.   ,  0.   ,  1.   ,  0.   ,  0.5  ,  0.   ,  6.   ,  0.   ],
       [ 0.   ,  1.   ,  0.   ,  0.   , -0.333,  0.333,  2.   ,  0.   ]])

HTML Tables

HTML tables allow web developers to arrange data into rows and columns.

<a href="https://www.w3schools.com/html/html_tables.asp">HTML Tables</a>


In [ ]:
#Reading the following cell in order to obtain some infromation related to the table in html

<h2>Simplex Tableau</h2>
<p>Setp 0:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> -3.00  </td><td style="text-align:center"> -5.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 4.00  </td><td style="text-align:center">-- </td></tr> 
<tr><td style="text-align:center">$x_4$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 2.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 12.00  </td><td style="text-align:center"> 6.00  </td></tr> 
<tr><td style="text-align:center">$x_5$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 3.00  </td><td style="text-align:center"> 2.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 18.00  </td><td style="text-align:center"> 9.00  </td></tr> 
</tbody></table>
<ul>
<li>
The value of reduced cost of the variable $x_2$ is -5.
</li>
<li>
The value of minimum ratio test is the minimum of {--, 6, 9} = 6.
</li>
<li>
Hence the variable $x_2$ should leave the basis.
</li>
 
 
 
</ul>
<p>Setp 1:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> -3.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 2.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 30.00  </td><td style="text-align:center"> </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 4.00  </td><td style="text-align:center"> 4.00  </td></tr> 
<tr><td style="text-align:center">$x_2$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 6.00  </td><td style="text-align:center">-- </td></tr> 
<tr><td style="text-align:center">$x_5$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 3.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> -1.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 6.00  </td><td style="text-align:center"> 2.00  </td></tr> 
</tbody></table>
<ul>
<li>
The value of reduced cost of the variable $x_1$ is -3.
</li>
<li>
The value of minimum ratio test is the minimum of {4, --, 2} = 2.
</li>
<li>
Hence the variable $x_1$ should leave the basis.
</li>
 
 
 
</ul>
<p>Setp 2:</p>
<table class="table table-bordered table-hover "><tbody><tr>
<th></th><th style="text-align:center">Z </th><th style="text-align:center">$x_1$ </th><th style="text-align:center">$x_2$ </th><th style="text-align:center">$x_3$ </th><th style="text-align:center">$x_4$ </th><th style="text-align:center">$x_5$ </th><th style="text-align:center">RHS </th><th style="text-align:center"> </th></tr> 
<tr><td style="text-align:center">Z </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.50  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 36.00  </td></tr> 
<tr><td style="text-align:center">$x_3$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.33  </td><td style="text-align:center"> -0.33  </td><td style="text-align:center"> 2.00  </td></tr> 
<tr><td style="text-align:center">$x_2$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.50  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 6.00  </td></tr> 
<tr><td style="text-align:center">$x_1$ </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 1.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> 0.00  </td><td style="text-align:center"> -0.33  </td><td style="text-align:center"> 0.33  </td><td style="text-align:center"> 2.00  </td></tr> 
</tbody></table>
<ul>
<li>
The reduced cost vector is nonnegative.</li>
<li>
The optimal solution is ($x_1$, $x_2$, $x_3$, $x_4$, $x_5$) = (2.00, 6.00, 2.00, 0, 0)
</li>
</ul>

<a href="http://140.116.53.170/classes/numericalOptimization/simplexMethods/simplexTableau.html">Simplex Tableau Example</a>

In [8]:
f = open("SimplexTabuleau.html", "w")
f.write("Now the file has more content!")
f.close()

In [1]:
f = open("SimplexTabuleau.html", "w")
f.write("Now the file has more content!")

f.write("<table>")
f.write("<tr>")
f.write("<th>")
f.write(str(A[1,1]))
f.write("</th>")
f.write("    <th>Contact</th>")
f.write("    <th>Country</th>")
f.write("  </tr>")
f.write("</table>")

f.close()

NameError: name 'A' is not defined

Open file and return a corresponding file object. If the file cannot be opened, an OSError is raised. See Reading and Writing Files for more examples of how to use this function.

<a href="https://docs.python.org/3/library/functions.html#open">open()</a>

<a href="https://docs.python.org/3/tutorial/inputoutput.html#tut-files">Reading and Writing Files</a>



# Exercise:

Economical Diet Proposal:
<ol>
<li> There are available 4 different foods at the market.</li>
<li> There are 3 basic nutritional ingredients.</li>
<li> In order to achieve a balanced diet, each individual must receive at least 800g, 90g, 50g of the first, second, third nutrient per day.</li>
<li> The first, second, third, forth food are sold at prices 200NTD, 300NTD, 400NTD, 250NTD per unit.</li>
<li> The relation between food and nutrient is describes as follows.</li>
<li> Determine the most economical diet that satisfies the minimum nutritional requirements.</li>
</ol>

<table class="table table-bordered table-hover "><tbody>
    <tr>
        <td style="text-align:center"> </td>
        <td style="text-align:center">1 </td>
        <td style="text-align:center">2 </td>
        <td style="text-align:center">3 </td>
        <td style="text-align:center">4 </td>
        <td style="text-align:center">minimum requirements </td>    
    </tr>
    <tr>
        <td style="text-align:center">1</td>
        <td style="text-align:center">0.95 </td>
        <td style="text-align:center">0.85 </td>
        <td style="text-align:center">0.75 </td>
        <td style="text-align:center">0.99 </td>
        <td style="text-align:center">800 </td>    
    </tr>
    <tr>
        <td style="text-align:center">2</td>
        <td style="text-align:center">0.025 </td>
        <td style="text-align:center">0.1 </td>
        <td style="text-align:center">0.2 </td>
        <td style="text-align:center">0.01 </td>
        <td style="text-align:center">90</td>    
    </tr>
    <tr>
        <td style="text-align:center">3</td>
        <td style="text-align:center">0.025 </td>
        <td style="text-align:center">0.15 </td>
        <td style="text-align:center">0.05 </td>
        <td style="text-align:center">0 </td>
        <td style="text-align:center">50 </td>    
    </tr>
    <tr>
        <td style="text-align:center">Price of food </td>
        <td style="text-align:center">200</td>
        <td style="text-align:center">300</td>
        <td style="text-align:center">400</td>
        <td style="text-align:center">250</td>
        <td style="text-align:center"> </td>    
    </tr>
</tbody></table>

$$
\begin{array}{rl}
\min & 200x_1+300x_2+400x_3+250x_4\\
s.t. & 0.95x_1+0.85x_2+0.75x_3+0.99x_4\geq 800\\
~& 0.025x_1+0.1x_2+0.2x_3+0.01x_4\geq 90\\
~& 0.025x_1+0.15x_2+0.05x_3\geq 50\\
~& x_1\geq 0,~x_2\geq 0,~x_3\geq 0,~x_4\geq 0\\
\end{array}
$$

$$
\begin{array}{rl}
\max & -200x_1-300x_2-400x_3-250x_4\\
s.t. & 0.95x_1+0.85x_2+0.75x_3+0.99x_4\geq 800\\
~& 0.025x_1+0.1x_2+0.2x_3+0.01x_4\geq 90\\
~& 0.025x_1+0.15x_2+0.05x_3\geq 50\\
~& x_1\geq 0,~x_2\geq 0,~x_3\geq 0,~x_4\geq 0\\
\end{array}
$$

$$
\begin{array}{rl}
\max & -200x_1-300x_2-400x_3-250x_4\\
s.t. & 0.95x_1+0.85x_2+0.75x_3+0.99x_4-s_1= 800\\
~& 0.025x_1+0.1x_2+0.2x_3+0.01x_4-s_2= 90\\
~& 0.025x_1+0.15x_2+0.05x_3-s_3= 50\\
~& x_1\geq 0,~x_2\geq 0,~x_3\geq 0,~x_4\geq 0\\
\end{array}
$$